File: libs/monobitmap-manager/src/main/kotlin/mono/bitmap/manager/factory/TextBitmapFactory.kt
package mono.bitmap.manager.factory

import mono.graphics.bitmap.MonoBitmap
import mono.graphics.geo.Size
import mono.shape.extra.TextExtra
import mono.shape.extra.style.TextAlign

object TextBitmapFactory {
    fun toBitmap(
        boundSize: Size,
        renderableText: List<String>,
        extra: TextExtra,
        isTextEditingMode: Boolean
    ): MonoBitmap {
        val bgBitmap =
            RectangleBitmapFactory.toBitmap(boundSize, extra.boundExtra)
        val bitmapBuilder = MonoBitmap.Builder(boundSize.width, boundSize.height)
        if (!(boundSize.width == 1 && boundSize.height == 1 && isTextEditingMode)) {
            bitmapBuilder.fill(0, 0, bgBitmap)
        }

        val adjustedRenderableText = if (!isTextEditingMode) renderableText else emptyList()
        bitmapBuilder.fillText(adjustedRenderableText, boundSize, extra)
        return bitmapBuilder.toBitmap()
    }

    private fun MonoBitmap.Builder.fillText(
        renderableText: List<String>,
        boundSize: Size,
        extra: TextExtra
    ) {
        val hasBorder = extra.hasBorder()
        val rowOffset = if (hasBorder) 1 else 0
        val colOffset = if (hasBorder) 1 else 0

        val maxTextWidth = boundSize.width - colOffset * 2
        val maxTextHeight = (boundSize.height - rowOffset * 2).coerceAtLeast(0)

        val row0 = when (extra.textAlign.verticalAlign) {
            TextAlign.VerticalAlign.TOP -> rowOffset
            TextAlign.VerticalAlign.MIDDLE ->
                if (maxTextHeight < renderableText.size) {
                    rowOffset
                } else {
                    (maxTextHeight - renderableText.size) / 2 + rowOffset
                }
            TextAlign.VerticalAlign.BOTTOM ->
                if (maxTextHeight < renderableText.size) {
                    rowOffset
                } else {
                    maxTextHeight - renderableText.size + rowOffset
                }
        }

        val horizontalAlign = extra.textAlign.horizontalAlign
        for (rowIndex in renderableText.indices.take(maxTextHeight)) {
            val row = renderableText[rowIndex]
            val col0 = when (horizontalAlign) {
                TextAlign.HorizontalAlign.LEFT -> colOffset
                TextAlign.HorizontalAlign.MIDDLE -> (maxTextWidth - row.length) / 2 + colOffset
                TextAlign.HorizontalAlign.RIGHT -> maxTextWidth - row.length + colOffset
            }
            for (colIndex in row.indices) {
                val char = row[colIndex]
                if (char != ' ') {
                    put(row0 + rowIndex, col0 + colIndex, char)
                }
            }
        }
    }
}

File: monosketch-svelte/src/lib/mono/monobitmap/manager/text-bitmap-factory.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { RectangleBitmapFactory } from './rectangle-bitmap-factory';
import type { Size } from '$libs/graphics-geo/size';
import { MonoBitmap } from '$mono/monobitmap/bitmap/monobitmap';
import type { TextExtra } from '$mono/shape/extra/shape-extra';
import { TextHorizontalAlign, TextVerticalAlign } from '$mono/shape/extra/style';

/**
 * A factory class to create a bitmap for a text shape.
 */
export class TextBitmapFactory {
    static toBitmap(
        boundSize: Size,
        renderableText: string[],
        extra: TextExtra,
        isTextEditingMode: boolean,
    ): MonoBitmap.Bitmap {
        const bgBitmap = RectangleBitmapFactory.toBitmap(boundSize, extra.boundExtra);
        const bitmapBuilder = new MonoBitmap.Builder(boundSize.width, boundSize.height);
        if (!(boundSize.width === 1 && boundSize.height === 1 && isTextEditingMode)) {
            bitmapBuilder.fillBitmap(0, 0, bgBitmap);
        }

        const adjustedRenderableText = isTextEditingMode ? [] : renderableText;
        this.fillText(bitmapBuilder, adjustedRenderableText, boundSize, extra);
        return bitmapBuilder.toBitmap();
    }

    private static fillText(
        bitmapBuilder: MonoBitmap.Builder,
        renderableText: string[],
        boundSize: Size,
        extra: TextExtra,
    ) {
        const hasBorder = extra.hasBorder();
        const rowOffset = hasBorder ? 1 : 0;
        const colOffset = hasBorder ? 1 : 0;

        const maxTextWidth = boundSize.width - colOffset * 2;
        const maxTextHeight = Math.max(0, boundSize.height - rowOffset * 2);

        const row0 = this.getRow0(extra, rowOffset, maxTextHeight, renderableText);

        const horizontalAlign = extra.textAlign.horizontalAlign;
        const maxRowIndex = Math.min(renderableText.length, maxTextHeight);
        for (let rowIndex = 0; rowIndex < maxRowIndex; rowIndex++) {
            const row = renderableText[rowIndex];
            const col0 = this.getCol0(horizontalAlign, colOffset, maxTextWidth, row);
            for (let colIndex = 0; colIndex < row.length; colIndex++) {
                const char = row[colIndex];
                if (char !== ' ') {
                    bitmapBuilder.put(row0 + rowIndex, col0 + colIndex, char, char);
                }
            }
        }
    }

    private static getCol0(horizontalAlign: TextHorizontalAlign, colOffset: number, maxTextWidth: number, row: string) {
        switch (horizontalAlign) {
            case TextHorizontalAlign.LEFT:
                return colOffset;
            case TextHorizontalAlign.MIDDLE:
                return Math.floor((maxTextWidth - row.length) / 2) + colOffset;
            case TextHorizontalAlign.RIGHT:
                return maxTextWidth - row.length + colOffset;
        }
    }

    private static getRow0(extra: TextExtra, rowOffset: number, maxTextHeight: number, renderableText: string[]) {
        switch (extra.textAlign.verticalAlign) {
            case TextVerticalAlign.TOP:
                return rowOffset;
            case TextVerticalAlign.MIDDLE:
                return maxTextHeight < renderableText.length
                    ? rowOffset
                    : Math.floor((maxTextHeight - renderableText.length) / 2) + rowOffset;
            case TextVerticalAlign.BOTTOM:
                return maxTextHeight < renderableText.length
                    ? rowOffset
                    : maxTextHeight - renderableText.length + rowOffset;
        }
    };
}
